content/en/blog/_posts/2022-10-30-forensic-container-checkpointing/index.md

```diff
@@ -0,0 +1,175 @@
+
+```
+# curl -X POST "https://localhost:10250/checkpoint/default/counters/counter"
+```
+
+For completeness the following `curl` command-line options are necessary to
+create a checkpoint:
+
+```
+-k --cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key
+```
+
+Triggering this **kubelet** API will request the creation of a checkpoint from
+CRI-O and CRI-O will request a checkpoint from *runc* and *runc* will tell CRIU
+to do the actual checkpointing.
+
+Once the checkpointing has finished the checkpoint should be available at
+`/var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar`
+
+This tar archive can now be used to restore the container somewhere else.
+
+#### Restore checkpointed container outside of Kubernetes
+
+With the checkpoint tar archive it is possible to restore the container outside
+of Kubernetes in a sandboxed instance of CRI-O. For better user experience
+during restore it is recommended to use the latest version of CRI-O from the
+*main* CRI-O GitHub branch. Using the CRI-O release v1.25 requires the user to
+manually create certain directories Kubernetes would create before starting the
+container.
+
+The first step to restore a container outside of Kubernetes is to create a pod
+using *crictl*: `crictl runp pod-config.json`
+
+Then the checkpointed container can be restored into the just create pod:
+`crictl create <POD_ID> container-config.json pod-config.json`
+
+Instead of specifying a container image in a registry in `container-config.json`
+it is necessary to specify the checkpoint archive as created above:
+```
+{
+  "metadata": {
+      "name": "counter"
+  },
+  "image":{
+      "image": "/var/lib/kubelet/checkpoints/<checkpoint-archive>.tar"
+  }
+}
+```
+
+Now the container needs to be started using `crictl start <CONTAINER_ID>` and a
+copy of the previously checkpointed container should be running.
+
+#### Restore checkpointed container in Kubernetes
+
+To restore the previously checkpointed container directly in Kubernetes it is
+necessary to convert the checkpoint archive into an image that can be pushed to
+a registry.
+
+One possible way to convert the local checkpoint archive consists of the
+following steps with the help of *buildah*:
+```
+# newcontainer=$(buildah from scratch)
+# buildah add $newcontainer /var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar /
+# buildah config --annotation=io.kubernetes.cri-o.annotations.checkpoint.name=<container-name> $newcontainer
+# buildah commit $newcontainer checkpoint-image:latest
+```
+
+This image now needs to be a pushed to a registry:
+```
+# podman push localhost/checkpoint-image:latest registry/user/checkpoint-image:latest
+```
+
+To restore this container (`registry/user/checkpoint-image:latest`) the
+image needs to be listed in the pod definition:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: <pod-name>
+spec:
```

Comment:
I cannot answer that. No idea. Sorry. I never used a `nodeName` during testing so that I know it works without it.

Reply:
If we want to restore the Pod to a specific Node, we should specify where to place the new Pod. Otherwise, the scheduler will pick one and that might not match our expectation.

Reply:
Thanks. Makes sense. I will add `nodeName`.